ENH: Make the numpy advanced iterator example more elaborate

seberg opened the pull request:
This uses the iterator to allocate the output. It also adds some
checks so that the code will not fail on non-native byte order
or unalinged inputs.
It also makes use of the inner loop optimization which should be
used for most iterations such as here.

Possibly this goes too far and a simpler version is better?

Closes gh-402.

---

It should probably be good, but I should go over it one more time myself before merging. I am not sure if this may be adding a bit too many bells and whistles for a simple tutorial.

pdebuyl commented:
The past example was not "hello world" neither, so I see no issue with the level of complexity.

I had a look and tested the code with Python 2.7, 3.6 and 3.7.

Did you want to check this further or can I merge?

seberg commented:
@pdebuyl went through it, just changed the comments/whitespace a bit. I think it should be OK, the question is if it is didactically nice.

I added more involved tests for the corner cases to the test file, not sure if this is something you like or if it would have the right style.

pdebuyl commented:
Hi @seberg 

Sorry for the (even for me) unusually long delay. As I am not familar with the NumPy C API it took me a few times to get to the code. A lot of the boilerplate is standard, which is good :-)

I have one issue though. The numerical results are not correct. I printed y for completeness:

```
[ 1.          0.1015857   0.1015857  -0.16299078  0.1015857   1.
 -0.16299078 -0.78223089  0.1015857   0.7421542   1.          0.1015857
 -0.16299078  0.7421542  -0.78223089  1.          0.1015857  -0.16299078
  0.7421542  -0.78223089  1.          0.7421542   0.1015857  -0.78223089
 -0.16299078  1.          0.7421542   0.1015857  -0.78223089 -0.16299078
  1.          0.7421542   0.1015857  -0.78223089 -0.16299078  1.          0.7421542
  0.1015857  -0.78223089 -0.16299078  1.         -0.78223089  0.7421542
  0.7421542   0.1015857   1.         -0.78223089  0.7421542  -0.16299078
  0.1015857   1.         -0.78223089  0.7421542  -0.16299078  0.1015857   1.
 -0.78223089  0.7421542  -0.16299078  0.1015857   1.         -0.78223089
  0.7421542 ]
```

I can't find where the error comes from. I could reproduce on two linux computers (ubuntu 16.04 and debian sid), Python 2.7, 3.6 and 3.7.

Looking at the documentation, I can't see what is wrong :-(

seberg commented:
Thanks for having a look. I added an `assert()` for the out stride (that using `out++` is indeed fine).

Other than that, the issue was that I wrote `(double) *in` instead of `*(double *)in`...

pdebuyl commented:
Thank you very much for the fix!
